url56b1dc76d3_wufunxiang_Libs_tgz-pJ8-org_msgpack_util_TemplatePrecompilerJ8/src/org/msgpack/MessagePack.java

```diff
@@ -0,0 +1,837 @@
+     * @since 0.6.0
+     * @param v
+     * @param template
+     * @return
+     * @throws IOException
+     */
+    public <T> byte[] write(T v, Template<T> template) throws IOException {
+        BufferPacker pk = createBufferPacker();
+        template.write(pk, v);
+        return pk.toByteArray();
+    }
+
+    /**
+     * Serializes specified object to output stream.
+     * 
+     * @since 0.6.0
+     * @param out
+     *            output stream
+     * @param v
+     *            serialized object
+     * @throws IOException
+     */
+    public <T> void write(OutputStream out, T v) throws IOException {
+        Packer pk = createPacker(out);
+        if (v == null) {
+            pk.writeNil();
+        } else {
+            @SuppressWarnings("unchecked")
+            Template<T> tmpl = registry.lookup(v.getClass());
+            tmpl.write(pk, v);
+        }
+    }
+
+    /**
+     * Serializes object to output stream by specified template.
+     * 
+     * @since 0.6.0
+     * @param out
+     *            output stream
+     * @param v
+     *            serialized object
+     * @param template
+     *            serializer/deserializer for the object
+     * @throws IOException
+     */
+    public <T> void write(OutputStream out, T v, Template<T> template)
+            throws IOException {
+        Packer pk = createPacker(out);
+        template.write(pk, v);
+    }
+
+    /**
+     * Serializes {@link org.msgpack.type.Value} object to byte array.
+     * 
+     * @since 0.6.0
+     * @param v
+     *            serialized {@link org.msgpack.type.Value} object
+     * @return output byte array
+     * @throws IOException
+     */
+    public byte[] write(Value v) throws IOException {
+        // FIXME ValueTemplate should do this
+        BufferPacker pk = createBufferPacker();
+        pk.write(v);
+        return pk.toByteArray();
+    }
+
+    /**
+     * Deserializes specified byte array to {@link org.msgpack.type.Value}
+     * object.
+     * 
+     * @since 0.6.0
+     * @param bytes
+     *            input byte array
+     * @return
+     * @throws IOException
+     */
+    public Value read(byte[] bytes) throws IOException {
+        return read(bytes, 0, bytes.length);
+    }
+
+    /**
+     * Deserializes byte array to {@link org.msgpack.type.Value} object.
+     * 
+     * @since 0.6.0
+     * @param bytes
+     * @param off
+     * @param len
+     * @return
+     * @throws IOException
+     */
+    public Value read(byte[] bytes, int off, int len) throws IOException {
+        return createBufferUnpacker(bytes, off, len).readValue();
+    }
+
+    /**
+     * Deserializes {@link java.nio.ByteBuffer} object to
+     * {@link org.msgpack.type.Value} object.
+     * 
+     * @since 0.6.0
+     * @param buffer
+     *            input buffer
+     * @return
+     * @throws IOException
+     */
+    public Value read(ByteBuffer buffer) throws IOException {
+        return createBufferUnpacker(buffer).readValue();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the resource returned by the *createBufferUnpacker* method in a try-finally block. Execution paths that do not contain closure statements were detected.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).